File: api/neck/Models/Chord.cs
using neck.Enums;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;
using System.Diagnostics;
using System.Linq;
using static neck.Enums.ChordEnums;

namespace neck.Models
{
	public class Chord : DbEntity
	{
		private List<Key> _keys;
		private List<Note> _tones;

		public Guid RootId { get; set; }
		public Note Root { get; set; }

		public ChordModifier Modifier { get; set; }

		public string Label => $"{Root?.Label} {Modifier}";

		[NotMapped]
		public List<Key> Keys => _keys;

		[NotMapped]
		public List<Note> Tones
		{
			get
			{
				if (_tones == null && Root != null)
				{
					_tones = mapComponents(Root, Modifier);
				}

				return _tones;
			}
		}

		public Chord() { }

		public Chord(Note root, ChordModifier modifier)
		{
			Root = root;
			Modifier = modifier;

			_keys = new List<Key>
			{
				new Key(Root, Modifier == ChordModifier.Minor ? KeyType.Minor : KeyType.Major)
			};

			_tones = mapComponents(Root, Modifier);
		}

		//	MAJOR
		//
		//  C           D           E     F           G           A           B   
		//        C#/Db       D#/Eb             F#/Gb       G#/Ab       A#/Bb
		//  B#                      Fb    E#                                  Cb
		//  0     1     2     3     4     5     6     7     8     9     10    11 

		//	MINOR
		//
		//	

		// I        tonic
		// ii       supertonic
		// iii      mediant
		// IV       subdominant
		// V        dominant
		// vi	      submediant
		// viio     leading tone (#)
		//            / subtonic (b)
		//  / ♭VII

		//  I   ii  iii IV  V   vi  viio
		//  ----------------------------
		//  C   Dm  Em  F   G   Am  Bdim

		private List<Interval> getIntervals(ChordModifier mod)
		{
			// TODO: Currently exist on front end
			switch (mod)
			{
				case ChordModifier.Major:
					return new List<Interval> { Interval.Root, Interval.MajorThird, Interval.PerfectFifth };
				case ChordModifier.Minor:
					return new List<Interval> { Interval.Root, Interval.MinorThird, Interval.PerfectFifth };
				case ChordModifier.Diminished:
					return new List<Interval> { Interval.Root, Interval.MinorThird, Interval.DiminishedFifth };
				case ChordModifier.MajorSeventh:
					return new List<Interval> { Interval.Root, Interval.MajorThird, Interval.PerfectFifth, Interval.MajorSeventh };
				case ChordModifier.MinorSeventh:
					return new List<Interval> { Interval.Root, Interval.MinorThird, Interval.PerfectFifth, Interval.MinorSeventh };
				case ChordModifier.DominantSeventh:
					return new List<Interval> { Interval.Root, Interval.MajorThird, Interval.PerfectFifth, Interval.MinorSeventh };
				case ChordModifier.SuspendedFour:
					throw new Exception("Not implemented");
				case ChordModifier.Augmented:
					return new List<Interval> { Interval.Root, Interval.MajorThird, (Interval)AugmentedInterval.AugmentedFifth };
				case ChordModifier.AugmentedSeventh:
					return new List<Interval> { Interval.Root, Interval.MajorThird, (Interval)AugmentedInterval.AugmentedFifth, Interval.MinorSeventh };
				default:
					return null;
			}
		}

		private List<Note> mapComponents(Note root, ChordModifier modifier)
		{

			var intervals = getIntervals(modifier);

			var scale = new Scale(root, getMode(modifier));

			List<Note> components;
			if (modifier == ChordModifier.Augmented || modifier == ChordModifier.AugmentedSeventh)
			{
				components = intervals
					.Select(i =>
					{
						var note = scale.Notes.FirstOrDefault(n => n.Interval == i);
						if (note == null)
						{
							var flat = scale.Notes.FirstOrDefault(n => n.Interval == i + 1);
							if (flat != null && (flat.Interval == Interval.PerfectFifth || flat.Interval == Interval.MajorSeventh))
							{
								return new Note(flat.Base, NoteSuffix.Flat);
							}
							var sharp = scale.Notes.FirstOrDefault(n => n.Interval == i - 1);
							if (sharp != null && (sharp.Interval == Interval.PerfectFifth || sharp.Interval == Interval.MajorSeventh))
							{
								return new Note(sharp.Base, NoteSuffix.Sharp);
							}
						}

						return note;
					})
					.ToList();
			}
			else
			{
				components = intervals.Select(i => scale.Notes.FirstOrDefault(n => n.Interval == i)).ToList();
			}

			return components;
		}

		private Mode getMode(ChordModifier mod)
		{
			switch (mod)
			{
				case ChordModifier.Minor:
				case ChordModifier.MinorSeventh:
					return new Mode(ModeType.Aeolian);
				case ChordModifier.DominantSeventh:
					return new Mode(ModeType.Mixolydian);
				default:
					return new Mode(ModeType.Ionian);
			}
		}

	}
}

File: web/src/models/Chord.ts
import { filter, map } from "lodash";

import { Key, Mode, Note, Scale } from ".";
import { ApiEntity } from "../network/ApiEntity";

export enum ChordModifier {
  Major,
  Minor,
  Diminished,
  Augmented,
  SuspendedSecond,
  SuspendedFourth,
  MajorSeventh,
  MinorSeventh,
  DominantSeventh,
  DiminishedSeventh,
  AugmentedSeventh,
  // MajorSixth,
  // MinorSixth,
}

// I        tonic
// ii       supertonic
// iii      mediant
// IV       subdominant
// V        dominant
// vi	      submediant
// viio     leading tone (#)
//            / subtonic (b)
//  / ♭VII

//  I   ii  iii IV  V   vi  viio
//  ----------------------------
//  C   Dm  Em  F   G   Am  Bdim

// const getPitches = (mod: ChordModifier): number[] => {
//   switch (mod) {
//     case ChordModifier.Major:
//       return [0, 2, 4];
//     // case ChordModifier.DominantSeventh:
//     //   return [0, 2, 4, 7];
//   }

//   return [];
// }

// const calculatePitches = (scale: Scale, mod: ChordModifier): Note[] => {
//   return map(getPitches(mod), p => {
//     const degree = p % scale.Notes.length;
//     const noteResult = filter(scale.Notes, n => n.Degree == degree);
//     return noteResult[0];
//   });
// }

export class Chord extends ApiEntity {

  public Root: Note;
  public Modifier: ChordModifier;

  public Keys: Key[];

  public Degree: number;

  public Tones: Note[];

  constructor(root: Note, mod: ChordModifier) {
    super();

    this.Root = root;
    this.Modifier = mod || ChordModifier.Major;

    // this.Keys = this.Modifier === ChordModifier.Minor
    //   ? [new Key(this.Root).RelativeMajor]
    //   : [new Key(this.Root)];
  }

  get Label(): string {
    return `${this.Root.Label} ${Chord.getModifierLabel(this.Modifier)}`
  }

  get Abbreviated(): string {
    return `${this.Root.Label}${this.Modifier}`;
  }

  get ModifierLabel(): string {
    return Chord.getModifierLabel(this.Modifier);
  }

  // TODO: Figure out where these should go
  static getModifierLabel = (mod: ChordModifier): any => {
    switch (mod) {
      case ChordModifier.Major:
        return 'Major';
      case ChordModifier.Minor:
        return 'Minor';
      case ChordModifier.Diminished:
        return 'Dim';
      case ChordModifier.Augmented:
        return 'Aug';
      case ChordModifier.SuspendedSecond:
        return 'Sus2';
      case ChordModifier.SuspendedFourth:
        return 'Sus4';
      case ChordModifier.MajorSeventh:
        return 'Maj7';
      case ChordModifier.MinorSeventh:
        return 'Min7';
      case ChordModifier.DominantSeventh:
        return '7';
      case ChordModifier.DiminishedSeventh:
        return 'Dim7';
      case ChordModifier.AugmentedSeventh:
        return 'Aug7';
    }
  }

  static getModifierAbbreviation = (mod: ChordModifier): any => {
    switch (mod) {
      case ChordModifier.Major:
        return 'maj';
      case ChordModifier.Minor:
        return 'min';
      case ChordModifier.Diminished:
        return 'dim';
      case ChordModifier.Augmented:
        return 'aug';
      case ChordModifier.SuspendedSecond:
        return 'sus2';
      case ChordModifier.SuspendedFourth:
        return 'sus4';
      case ChordModifier.MajorSeventh:
        return 'maj7';
      case ChordModifier.MinorSeventh:
        return 'min7';
      case ChordModifier.DominantSeventh:
        return '7';
      case ChordModifier.DiminishedSeventh:
        return 'dim7';
      case ChordModifier.AugmentedSeventh:
        return 'aug7';
    }
  }
}